prolog.cl

```diff
@@ -841,6 +842,747 @@
+      '("Pipe"
+      "Internal table stage"
+      "Internal named stage")
+      
+      '("Serverless task"
+      "Internal user stage"
+      "Storage integration"
+      )
+)
+
+  (multi "You are the owner of a table T1 which is in schema S1. The schema is in database D1. In order to grant read-only permissions of this table to a newly created role R1, you will need to…(select all that apply)"
+         6
+        '("Grant ‘USAGE’ on database D1"
+          "Grant ‘USAGE’ on schema S1"
+          "Grant ‘SELECT’ on table T1")
+        
+         '("Grant ‘SELECT’ on database D1"
+           "Grant ‘SELECT’ on schema S1"
+           "Grant ‘USAGE’ on table T1"
+           )
+)
+
+
+   (multi "Which are characteristics of Snowpipe"
+         6
+        '("Load history is stored for 14 days."
+        "Load history must be requested from Snowflake via a REST endpoint, SQL table function, or ACCOUNT_USAGE view."
+        "Loads are combined or split into a single or multiple transactions based on the number and size of the rows in each data file"
+        "Uses Snowflake-supplied compute resources.
+
+")
+         '("Load history is stored for 64 days."
+           "Load history is available upon completion of the COPY statement as the statement output."
+           "Loads are always performed in a single transaction."
+           "Requires a user-specified warehouse to execute COPY statements."
+           )
+)
+
+   (multi "Which are characteristics of batch data loads"
+         6
+
+      '("Load history is stored for 64 days."
+           "Load history is available upon completion of the COPY statement as the statement output."
+           "Loads are always performed in a single transaction."
+           "Requires a user-specified warehouse to execute COPY statements."
+           )
+        '("Load history is stored for 14 days."
+        "Load history must be requested from Snowflake via a REST endpoint, SQL table function, or ACCOUNT_USAGE view."
+        "Loads are combined or split into a single or multiple transactions based on the number and size of the rows in each data file"
+        "Uses Snowflake-supplied compute resources."
+        )
+
+)
+   (fact "Clustering key consists of one or more table columns/expressions, which can be of any data type,"
+         '("except" "including")
+         " GEOGRAPHY, VARIANT, OBJECT, or ARRAY")
+
+
+   (fact "Clustering keys"
+         '("cannot" can"")
+         "be defined for hybrid tables.")
+
+  (multi "This column contains the Kafka message."
+         4
+        '("RECORD_CONTENT")
+        
+         '("RECORD_MESSAGE"
+           "RECORD_DATA"
+           "RECORD_METADATA"
+           )
+)
+
+  (multi "Which format(s) can be used for Kafka messages?"
+         4
+        '("Arvo"
+          "JSON")
+        
+         '("Parquet"
+           "CSV"
+           "ORC"
+           "XML"
+           )
+)
+
+
+(multi "What are characteristics of Snowpark Python packages? (Choose three.)"
```

Comment:
Remove the "choose three" instruction, since the question posing apparatus can/should be in charge of that.